### ISSUE TITLE
Compiling error with latest Xamarin.iOS

### ISSUE THREAD

**Issue by streinhard** [martijn00/LottieXamarin#199]
## 🐛 Bug Report

When using Lottie.Forms 2.7.0 with latest Xamarin.Forms, Microsoft Visual Studio for Mac and Xamarin.iOS, I get a compiling error:

Error MT2101: Can't resolve the reference 'System.Void ObjCRuntime.Runtime::ReleaseBlockOnMainThread(System.IntPtr)', referenced from the method 'System.Void ObjCRuntime.Trampolines/NIDLOTAnimationCompletionBlock::Finalize()' in 'Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065'. (MT2101)

Removing Lottie from the project makes the error go away.

When I first had this error, I was using these versions:

Visual Studio for Mac: 7.6.10
Xamarin.iOS: 12.0.0.15
XCode: 10.0

Since XCode was just updated recently, I thought it might be worth a try. Those Versions were:

Visual Studio for Mac: 7.6.11
Xamarin.iOS: 12.1.0.15
XCode: 10.1

However, the same error was still present. 

~~Then I tried to download and build LottieXamarin myself, using the created NuGets and et voila: Everything works. Maybe the Library needs to be compiled with updated versions of everything to support the latest Xamarin Releases.~~ See comment below.

### Expected behavior

LottieXamarin just works (tm) with latest Xamarin releases.

### Reproduction steps

1. Update Xamarin stuff
2. Create a new Xamarin.Forms Project
3. Install NuGet Com.Airbnb.Xamarin.Forms.Lottie in all projects
4. Add a simple AnimationView to a Page (Otherwise Linker might remove Lottie)
5. Build an run on an actual Device (Simulator is not affected)

### Configuration

**Version:** 2.7.0

**Platform:** 
- [ x ] :iphone: iOS
- [ ] :robot: Android
- [ ] :checkered_flag: WPF
- [ ] :earth_americas: UWP
- [ ] :apple: MacOS
- [ ] :tv: tvOS
- [ x ] :monkey: Xamarin.Forms

**Comment by martijn00** [comment]
Hmm interesting. I actually used the latest vs4mac beta to compile. Can you check if it works with that?

**Comment by streinhard** [comment]
Surprise surprise, it works when switching to vs4mac beta channel 😄 . Unfortunately not on stable channel, where probably most developers work.

In my description, I wrote that it works when I compile LottieXamarin myself. However, I could not reproduce that, so it probably doesn't have anything to do with XCode Versions. I probably just depends on the Xamarin.iOS Version. On current version 12.1.0.15, the bug can be reproduced when recompiling LottieXamarin.

Maybe it has to something with this Xamarin.iOS issue: https://github.com/xamarin/xamarin-macios/pull/4312. Not sure though.

**Comment by streinhard** [comment]
Here's a small repro project.

[XamarinLottie-Issue199.zip](https://github.com/martijn00/LottieXamarin/files/2554304/XamarinLottie-Issue199.zip)

**Comment by ianvink** [comment]
First of all, thank you for all your work and efforts. Wonderful tool.

I can confirm this also happens to me on Xamarin  iOS 12.1.0.15 with Lottie 2.7. Linker fails with this:


`
    Mono.Linker.MarkException: Error processing method: 'System.Void ObjCRuntime.Trampolines/NIDLOTAnimationCompletionBlock::Finalize()' in assembly: 'Lottie.iOS.dll' ---> Mono.Cecil.ResolutionException: Failed to resolve System.Void ObjCRuntime.Runtime::ReleaseBlockOnMainThread(System.IntPtr)
      at Mono.Linker.Steps.MarkStep.HandleUnresolvedMethod (Mono.Cecil.MethodReference reference) [0x0001b] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1669 
      at Mono.Linker.Steps.MarkStep.MarkMethod (Mono.Cecil.MethodReference reference) [0x00064] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1374 
      at Xamarin.Linker.Steps.MobileMarkStep.MarkMethod (Mono.Cecil.MethodReference reference) [0x00001] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/MobileMarkStep.cs:140 
      at Xamarin.Linker.Steps.CoreMarkStep.MarkMethod (Mono.Cecil.MethodReference reference) [0x00001] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/CoreMarkStep.cs:228 
      at MonoTouch.Tuner.MonoTouchMarkStep.MarkMethod (Mono.Cecil.MethodReference reference) [0x00001] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/MonoTouch.Tuner/MonoTouchMarkStep.cs:98 
      at Mono.Linker.Steps.MarkStep.MarkInstruction (Mono.Cecil.Cil.Instruction instruction) [0x00048] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1638 
      at Xamarin.Linker.Steps.CoreMarkStep.MarkInstruction (Mono.Cecil.Cil.Instruction instruction) [0x00042] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/CoreMarkStep.cs:223 
      at Mono.Linker.Steps.MarkStep.MarkMethodBody (Mono.Cecil.Cil.MethodBody body) [0x000a5] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1617 
      at MonoTouch.Tuner.MonoTouchMarkStep.MarkMethodBody (Mono.Cecil.Cil.MethodBody body) [0x0000d] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/MonoTouch.Tuner/MonoTouchMarkStep.cs:160 
      at Mono.Linker.Steps.MarkStep.ProcessMethod (Mono.Cecil.MethodDefinition method) [0x0019d] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1465 
      at Xamarin.Linker.Steps.CoreMarkStep.ProcessMethod (Mono.Cecil.MethodDefinition method) [0x00060] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/CoreMarkStep.cs:187 
      at Mono.Linker.Steps.MarkStep.ProcessQueue () [0x0001f] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:188 
       --- End of inner exception stack trace ---
      at Mono.Linker.Steps.MarkStep.ProcessQueue () [0x0002c] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:190 
      at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue () [0x00012] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:174 
      at Mono.Linker.Steps.MarkStep.Process () [0x00146] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:136 
      at Mono.Linker.Steps.MarkStep.Process (Mono.Linker.LinkContext context) [0x0000f] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:77 
      at Xamarin.Linker.Steps.MobileMarkStep.Process (Mono.Linker.LinkContext context) [0x00001] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/MobileMarkStep.cs:33 
      at Xamarin.Linker.Steps.CoreMarkStep.Process (Mono.Linker.LinkContext context) [0x00017] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/CoreMarkStep.cs:32 
      at MonoTouch.Tuner.MonoTouchMarkStep.Process (Mono.Linker.LinkContext context) [0x0001d] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/MonoTouch.Tuner/MonoTouchMarkStep.cs:36 
      at Mono.Linker.Pipeline.Process (Mono.Linker.LinkContext context) [0x0001f] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker/Pipeline.cs:127 
      at MonoTouch.Tuner.Linker.Process (MonoTouch.Tuner.LinkerOptions options, MonoTouch.Tuner.MonoTouchLinkContext& context, System.Collections.Generic.List`1[Mono.Cecil.AssemblyDefinition]& assemblies) [0x000e0] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Tuning.cs:80 
    ---
    --- inner exception
    Mono.Cecil.ResolutionException: Failed to resolve System.Void ObjCRuntime.Runtime::ReleaseBlockOnMainThread(System.IntPtr)
      at Mono.Linker.Steps.MarkStep.HandleUnresolvedMethod (Mono.Cecil.MethodReference reference) [0x0001b] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1669 
      at Mono.Linker.Steps.MarkStep.MarkMethod (Mono.Cecil.MethodReference reference) [0x00064] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1374 
      at Xamarin.Linker.Steps.MobileMarkStep.MarkMethod (Mono.Cecil.MethodReference reference) [0x00001] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/MobileMarkStep.cs:140 
      at Xamarin.Linker.Steps.CoreMarkStep.MarkMethod (Mono.Cecil.MethodReference reference) [0x00001] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/CoreMarkStep.cs:228 
      at MonoTouch.Tuner.MonoTouchMarkStep.MarkMethod (Mono.Cecil.MethodReference reference) [0x00001] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/MonoTouch.Tuner/MonoTouchMarkStep.cs:98 
      at Mono.Linker.Steps.MarkStep.MarkInstruction (Mono.Cecil.Cil.Instruction instruction) [0x00048] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1638 
      at Xamarin.Linker.Steps.CoreMarkStep.MarkInstruction (Mono.Cecil.Cil.Instruction instruction) [0x00042] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/CoreMarkStep.cs:223 
      at Mono.Linker.Steps.MarkStep.MarkMethodBody (Mono.Cecil.Cil.MethodBody body) [0x000a5] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1617 
      at MonoTouch.Tuner.MonoTouchMarkStep.MarkMethodBody (Mono.Cecil.Cil.MethodBody body) [0x0000d] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/MonoTouch.Tuner/MonoTouchMarkStep.cs:160 
      at Mono.Linker.Steps.MarkStep.ProcessMethod (Mono.Cecil.MethodDefinition method) [0x0019d] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:1465 
      at Xamarin.Linker.Steps.CoreMarkStep.ProcessMethod (Mono.Cecil.MethodDefinition method) [0x00060] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/linker/CoreMarkStep.cs:187 
      at Mono.Linker.Steps.MarkStep.ProcessQueue () [0x0001f] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/linker/linker/Linker.Steps/MarkStep.cs:188 
    ---
      at MonoTouch.Tuner.Linker.Process (MonoTouch.Tuner.LinkerOptions options, MonoTouch.Tuner.MonoTouchLinkContext& context, System.Collections.Generic.List`1[Mono.Cecil.AssemblyDefinition]& assemblies) [0x001f0] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Tuning.cs:99 
      at Xamarin.Bundler.Target.LinkAssemblies (System.Collections.Generic.List`1[Mono.Cecil.AssemblyDefinition]& assemblies, System.String output_dir, System.Collections.Generic.IEnumerable`1[T] sharedCodeTargets) [0x00256] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Target.cs:517 
      at Xamarin.Bundler.Target.ManagedLink () [0x00720] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Target.cs:664 
      at Xamarin.Bundler.Target.ProcessAssemblies () [0x000c2] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Target.cs:868 
      at Xamarin.Bundler.Application.ProcessAssemblies () [0x0002f] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Application.cs:1411 
      at Xamarin.Bundler.Application.BuildManaged () [0x00001] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Application.cs:839 
      at Xamarin.Bundler.Application+<>c.<BuildAll>b__140_1 (Xamarin.Bundler.Application v) [0x00000] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Application.cs:787 
      at System.Collections.Generic.List`1[T].ForEach (System.Action`1[T] action) [0x00024] in <71f4e3d7d65e42b0a699d02e59766d3c>:0 
      at Xamarin.Bundler.Application.BuildAll () [0x00050] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/Application.cs:787 
      at Xamarin.Bundler.Driver.Main2 (System.String[] args) [0x00423] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/mtouch.cs:1402 
      at Xamarin.Bundler.Driver.Main (System.String[] args) [0x00015] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/tools/mtouch/mtouch.cs:934 
Done building target "_CompileToNative" in project "Max.iOS.csproj" -- FAILED.
`

**Comment by ianvink** [comment]
Lottie 2.6.3 works though, I can Link it in Release mode on iOS with this setting. Update to Lottie 2.7 and the linker fails:

![build](https://user-images.githubusercontent.com/112916/48086457-c4156e80-e1b1-11e8-8c66-b567a3cab578.png)

**Comment by streinhard** [comment]
The reason for this issue might pretty well be a Bug in Xamarin.iOS itself.  The [error code spec in the docs](https://docs.microsoft.com/en-us/xamarin/ios/troubleshooting/mtouch-errors#MT2101) even asks for a bug report. I don't get the Mono Cecil part in the error log like @ianvink does, but this is probably  because I don't use any Fody Weavers in this project.

I will try again tomorrow to recompile LottieXamarin fresh and clean with current stable release of Xamarin.iOS, I could swear this made the error go away the first time I tried. If so, it might be a good idea to only use stable Xamarin releases to build this library.

**Comment by streinhard** [comment]
Okay, now I know why I couldn't reproduce the fix by rebuilding LottieXamarin with Xamarin stable channel (Transient NuGet dependencies, hooray). Also, I got sort of a clue what's going on and how this issue could be resolved.

TL; DR: Use Xamarin stable channel to build library packages

## What happened?

Xamarin.iOS had an issue with WKWebView which could crash your App: [#4130](https://github.com/xamarin/xamarin-macios/issues/4130)

In order to resolve this issue, they made sure all blocks ([those pesky Objective-C thingies](http://fuckingblocksyntax.com/)?) are released on the main thread. This resulted in PR [#4312](https://github.com/xamarin/xamarin-macios/pull/4312). As part of that fix, a new method `ReleaseBlockOnMainThread(IntPtr block)` was introduced to Xamarin.iOS runtime engine. The runtime enigne *somehow* makes sure this method gets called when a block is released.

#4312 got [merged into master](https://github.com/xamarin/xamarin-macios/commit/0a24bd603de792ba068f8fcd8a195443f11e68f5) and was released starting with Xamarin.iOS 12.2.1.9. This version is currently in Xamarin beta and alpha channel. When you build your Xamarin.iOS library with that version, your Assembly will end up with a reference to method `ReleaseBlockOnMainThread(IntPtr block)`.

Meanwhile, this method is not there yet in current stable Xamarin.iOS version 12.1.0.15, or any version before that. When the reference gets resolved during buildtime, the build fails because this method cannot be resolved.

## How does this affect LottieXamarin?

Since the latest LottieXamarin release was built using Xamarin.iOS 12.2.1.9, all native iOS assemblies have this issue. The affected NuGet packages are:

| Package | Version |
| -------- | -------- |
| Com.Airbnb.iOS.Lottie | 2.5.6 |
| Com.Airbnb.Xamarin.Forms.Lottie | 2.7.0 |

Previous versions are not affected:

| Package | Version |
| -------- | -------- |
| Com.Airbnb.iOS.Lottie | <= 2.5.4 |
| Com.Airbnb.Xamarin.Forms.Lottie | <= 2.6.3 |

The important part here is that basic iOS binding library also has the issue. When someone installs Xamarin.Forms.Lottie into a Xamarin.iOS package, he will also get the dependency to Com.Airbnb.iOS.Lottie and nuget will automatically download the latest version.

This means, in order to work around this issue, you have to:

 1. First install Com.Airbnb.iOS.Lottie 2.5.4
 2. Then install Com.Airbnb.Xamarin.Forms.Lottie 2.6.3

## How to fix this issue?

LottieXamarin needs to be rebuilt and released using Xamarin.iOS 12.1.0.15 (stable channel).

Was there a special reason to switch to beta channel? If so, you can just downgrade Xamarin.iOS and leave everything else in beta channel releases. If not, I suggest using only stable channel for library builds.

Hope this makes sense and I could help out.

Cheers and keep up the good work!

**Comment by ianvink** [comment]
Thank you for your excellent and complete investigation. Your contributions to the community are valuable.

**Comment by martijn00** [comment]
The new Xamarin version is in stable now. Update to that and it works.

**Comment by vincentcastagna** [comment]
@martijn00 Have you tested it ?  I'm still having this issue with latest Xamarin Forms 3.4.0.1008975 and Lottie 2.7.0 ... 

I had to follow @streinhard workaround to remove this error.

`Can't resolve the reference 'System.Void ObjCRuntime.Runtime::ReleaseBlockOnMainThread(System.IntPtr)', referenced from the method 'System.Void ObjCRuntime.Trampolines/NIDLOTAnimationCompletionBlock::Finalize()' in 'Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065'.	CognitivesTest.iOS	D:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets	794	`

**Comment by drznathou** [comment]
@Umar3x I encounter the same  exception  with latest Xamarin Forms 3.4.0.1008975 and Lottie 2.7.0

How ever @streinhard Solution works 

>  First install Com.Airbnb.iOS.Lottie 2.5.4
>  Then install Com.Airbnb.Xamarin.Forms.Lottie 2.6.3

**Comment by lalitthinker** [comment]
This is workin

> @Umar3x I encounter the same exception with latest Xamarin Forms 3.4.0.1008975 and Lottie 2.7.0
> 
> How ever @streinhard Solution works
> 
> > First install Com.Airbnb.iOS.Lottie 2.5.4
> > Then install Com.Airbnb.Xamarin.Forms.Lottie 2.6.3

this is working 👍 
I install new version of xamarin ios and got the same error.i will find solution after downgrade lottie.ios to 2.5.4

**Comment by aman720gupta** [comment]
@martijn00 I'm still facing compilation issue can someone help me with that:

` Error MT2101: Can't resolve the reference 'System.Void ObjCRuntime.Runtime::ReleaseBlockOnMainThread(System.IntPtr)', referenced from the method 'System.Void ObjCRuntime.Trampolines/NIDLOTAnimationCompletionBlock::Finalize()' in 'Xamarin.iOS, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065'. (MT2101)`

**Comment by markuspalme** [comment]
I can confirm that this is still an issue when performing a release/device build in Azure DevOps.

**Comment by martijn00** [comment]
Which version of Xamarin.iOS are you using in Azure or Appcenter?

**Comment by markuspalme** [comment]
The reference in my project file does not specify a version:
`<Reference Include="Xamarin.iOS" />`

I can't see a specific version on the log. AzureDevops still uses an old iOS SDK (11.4), Microsoft has promised to upgrade to 12.1 later this month. Could this be the cause of this issue?

**Comment by keozx** [comment]
This is happening to me also in Azure with Xamarin.iOS.framework 12.0.0.15 / Mono 5.12

**Comment by martijn00** [comment]
You need 12.1

**Comment by keozx** [comment]
> 
> 
> You need 12.1

Thanks it worked selecting the 12.2.1.15 which is latest on the macOS server as of now

**Comment by vishnu4** [comment]
I'm using Azure Devops and i'm getting this error.  How do i specify the latest framework with their Hosted macOS agent?

**Comment by markuspalme** [comment]
This can be fixed on Azure DevOps by adding a `bash` build step before the iOS build step (YAML):

```
- bash: |
   /bin/bash -c "sudo $AGENT_HOMEDIRECTORY/scripts/select-xamarin-sdk.sh 5_16_0_0"
   /bin/bash -c "echo '##vso[task.setvariable variable=MD_APPLE_SDK_ROOT;]'/Applications/Xcode_10.1.app;sudo xcode-select --switch /Applications/Xcode_10.1.app/Contents/Developer"
   
  displayName: 'Select SDK version - See https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=designer&viewFallbackFrom=azdevops#xamarin'

```

**Comment by david-chambers** [comment]
I'm seeing this issue manifest only in my Azure DevOps build.  @markuspalme bash script workaround resolved the issue for me - thank you!

project references:
```xml
<PackageReference Include="Xamarin.Forms" Version="4.2.0.848062" />
<PackageReference Include="Com.Airbnb.Xamarin.Forms.Lottie">
      <Version>3.0.2</Version>
    </PackageReference>
    <PackageReference Include="Com.Airbnb.iOS.Lottie">
      <Version>2.5.10</Version>
    </PackageReference>
```

azure build script

```yml
- job: iOS
  pool:
    vmImage: 'macOS-latest'
- task: XamariniOS@2
    inputs:
      solutionFile: '**/*iOS.csproj'
      configuration: '$(buildConfiguration)'
      buildForSimulator: false
      packageApp: true
```

Note: everything builds fine locally.  Is the iOS SDK version or mac OS version behind the fixed release still in azure devops?
